services/sqs/src/main/java/software/amazon/awssdk/services/sqs/internal/batchmanager/core/RequestBatchManager.java

```diff
@@ -0,0 +1,228 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.sqs.internal.batchmanager.core;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.services.sqs.batchmanager.BatchOverrideConfiguration;
+import software.amazon.awssdk.services.sqs.internal.batchmanager.core.BatchAndSend;
+import software.amazon.awssdk.services.sqs.internal.batchmanager.core.BatchConfiguration;
+import software.amazon.awssdk.services.sqs.internal.batchmanager.core.BatchKeyMapper;
+import software.amazon.awssdk.services.sqs.internal.batchmanager.core.BatchManager;
+import software.amazon.awssdk.services.sqs.internal.batchmanager.core.BatchResponseMapper;
+import software.amazon.awssdk.services.sqs.internal.batchmanager.core.BatchingExecutionContext;
+import software.amazon.awssdk.services.sqs.internal.batchmanager.core.BatchingMap;
+import software.amazon.awssdk.services.sqs.internal.batchmanager.core.IdentifiableMessage;
+import software.amazon.awssdk.utils.Validate;
+
+@SdkInternalApi
+public final class RequestBatchManager<RequestT, ResponseT, BatchResponseT> implements BatchManager<RequestT, ResponseT,
+    BatchResponseT> {
+    private final int maxBatchItems;
+    private final Duration maxBatchOpenInMs;
+
+    /**
+     * A nested map that keeps track of customer requests and the corresponding responses to be completed. Requests are batched
+     * together according to a batchKey that is calculated from the request by the service client.
+     */
+    private final BatchingMap<RequestT, ResponseT> requestsAndResponsesMaps;
+
+    /**
+     * Takes a list of identified requests in addition to a destination and batches the requests into a batch request.
+     * It then sends the batch request and returns a CompletableFuture of the response.
+     */
+    private final BatchAndSend<RequestT, BatchResponseT> batchFunction;
+
+    /**
+     * Unpacks the batch response, then transforms individual entries to the appropriate response type. Each entry's batch ID
+     * is mapped to the individual response entry.
+     */
+    private final BatchResponseMapper<BatchResponseT, ResponseT> responseMapper;
```

Comment:
Since we agreed we don't need to support batch managers generically, is this complexity necessary? e.g. do we need to abstract away the concept of batching and sending (BatchSend) or can we just implement all of that internally in this class?

Reply:
Yeap the old generic design doesnot hold for ReceiveMessage but it holds good for 3 other APIs like ChangeVisibility , SendMessage and DeleteMessage the generics are used here to handle these 3 APIs